src/Arcus.WebApi.Security/Authentication/CertificateAuthenticationValidator.cs

```diff
@@ -0,0 +1,252 @@
+        /// using the <paramref name="services"/> to retrieve registered instances that will provide the expected certificate values.
+        /// </summary>
+        /// <param name="clientCertificate">The client certificate to validate.</param>
+        /// <param name="requirements">The requirements to identify which parts of the <paramref name="clientCertificate"/> should be validated.</param>
+        /// <param name="services">The collection of registered services, (ex. from the request pipeline).</param>
+        /// <returns>
+        ///     <c>true</c> when the specified <paramref name="clientCertificate"/> is valid, <c>false</c> otherwise.
+        /// </returns>
+        /// <exception cref="ArgumentNullException">Thrown when the <paramref name="clientCertificate"/> is <c>null</c>.</exception>
+        /// <exception cref="ArgumentNullException">Thrown when the <paramref name="requirements"/> is <c>null</c>.</exception>
+        /// <exception cref="ArgumentNullException">Thrown when the <paramref name="services"/> is <c>null</c>.</exception>
+        /// <exception cref="ArgumentException">Thrown when the <paramref name="requirements"/> contains a configuration key that is <c>null</c>.</exception>
+        /// <exception cref="KeyNotFoundException">
+        ///     Thrown when the <paramref name="requirements"/> requires to validate a property of the <paramref name="clientCertificate"/>
+        ///     for which there exists no registered <see cref="IX509ValidationLocation"/> implementation.
+        /// </exception>
+        internal async Task<bool> ValidateCertificate(
+            X509Certificate2 clientCertificate,
+            IDictionary<X509ValidationRequirement, ConfiguredKey> requirements,
+            IServiceProvider services)
+        {
+            Guard.NotNull(clientCertificate, nameof(clientCertificate), "Certificate authentication validation requires a client certificate");
+            Guard.NotNull(requirements, nameof(requirements), "Certificate authentication validation requires a series of requirements and their configured key");
+            Guard.NotNull(services, nameof(services), "Certificate authentication validation requires a service object to retrieve registered services");
+            Guard.For<ArgumentException>(
+                () => requirements.Any(requirement => requirement.Value is null), 
+                "Certificate authentication requires all configured keys in the series of requirements to be not 'null'");
+
+            IDictionary<X509ValidationRequirement, (ConfiguredKey configuredKey, string expected)> requirementValues = 
+                await GetExpectedCertificateValues(requirements, services);
+            
+            ILogger logger = 
+                services.GetService<ILoggerFactory>()
+                       ?.CreateLogger<CertificateAuthenticationValidator>() 
+                ?? (ILogger) NullLogger.Instance;
+
+            return requirementValues.All(keyValue =>
```

Comment:
Same feedback, move lambda to a dedicated method which is called instead?

Reply:
Outdated...

Reply:
Not really - https://github.com/arcus-azure/arcus.webapi/pull/43/files#diff-a00c4d616f7db7b54040a476294f8c9dR52

Reply:
Ah, OK, I see.
Not sure if that would make it more readable? Now you see that every certificate requirement goes through the dedicated kind of validation in the exposed (internal) method.

Reply:
The reason why it would improve readability is that in the context of this method I don't care how the validation itself works, it just has to run it.

If I want to see the details, then I go to the method itself. 

In the main method it would just look like the following:
`return expectedValuesByRequirement.All(PerformValidation)`

Next to that, the list currently is fairly short but the more checks we add, the longer this inline method call will be.

Reply:
Thanks!